transport-native-io_uring/src/main/java/io/netty/channel/uring/RingBuffer.java

```diff
@@ -0,0 +1,37 @@
+/*
+ * Copyright 2020 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.channel.uring;
+
+import io.netty.util.internal.PlatformDependent;
+
+public class RingBuffer {
+  private final IOUringSubmissionQueue ioUringSubmissionQueue;
+  private final IOUringCompletionQueue ioUringCompletionQueue;
```

Comment:
Depending on where you want to take the design, this class _could_ grow to become a facade for both queues, so submitting tasks and taking completions happens here. That would make this class the central place where we know how many tasks have been queued up, how many have been submitted, and how many completions are waiting to be consumed. We could use this information to dynamically decide when and how often to call `io_uring_enter`.

It's not something you have to do now. Getting a simple integration up and running is a more important first step. Bt it's something to consider when you think about where to take the design.

Reply:
yeah good point 👍  it would be much more efficient to call less `io_uring_enter`, I'm really excited to try this flag `IOURING_SETUP_SQPOLL`, But I'm not sure if you can change it dynamically, probably not so easy